### PR TITLE
HDDS-9554. Introduce a CLI command to show the datanodes under decommissioning

### PR DESCRIPTION
**Pull request by Tejaskriya**
## What changes were proposed in this pull request?

This PR introduces a CLI command `ozone admin datanode decommissionStatus` in order to view the status of datanodes that are in DECOMMISSIONING and DECOMMISSIONED state.

Currently, monitoring decommissioning status from the CLI is done using ozone admin datanode list and checking the status of the nodes. If there are specific containers waiting to be replicated that are blocking decommissioning, the SCM log needs to be checked. This CLI command can display the necessary information.

Through this PR, the command displays the status of the datanode, associated pipelines and container counts when the decommissioning was started, and also the current state of the node, current pipelines and container counts, along with the start time of decommission.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-9554

## How was this patch tested?

Added unit tests in TestDecommissionStatusSubCommand

### CODE REVIEW

**Comment by Tejaskriya** [comment]
@errose28 @sodonnel Could you please help me review this PR?

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DecommissionStatusSubCommand.java [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+  private String uuid;
+
+  @Override
+  public void execute(ScmClient scmClient) throws IOException {
+    List<HddsProtos.Node> decommissioningNodes;
+
+    if (!Strings.isNullOrEmpty(uuid)) {
+      decommissioningNodes = scmClient.queryNode(DECOMMISSIONING, null,
+          HddsProtos.QueryScope.CLUSTER, "").stream().filter(p ->
+          p.getNodeID().getUuid().equals(uuid)).collect(Collectors.toList());
```
Comment (siddhantsangwan) [inline hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DecommissionStatusSubCommand.java:72]: Since we're only concerned with one datanode in this case, we can avoid asking the SCM for all datanodes that are decommissioning. Instead, let's just ask for that one datanode so we don't have to transfer not needed data over the network.

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DecommissionStatusSubCommand.java [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+      decommissioningNodes = scmClient.queryNode(DECOMMISSIONING, null,
+          HddsProtos.QueryScope.CLUSTER, "");
+      System.out.println("\nDecommission Status: DECOMMISSIONING - " +
+          decommissioningNodes.size() + " nodes");
+    }
+
+    Map<UUID, Map<HddsProtos.LifeCycleState, Long>> containerMap =
+        scmClient.getContainerMap();
+    Map<UUID, Integer> pipelineMap = scmClient.getPipelineMap();
+    List<Pipeline> currentPipelines = scmClient.listPipelines();
```
Comment (siddhantsangwan) [inline hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DecommissionStatusSubCommand.java:87]: We should consider what happens if any of these calls fails. Currently, I think an exception will be thrown to the user if any one call fails. Instead, I think we can handle each exception individually and show whatever information we successfully received, even if it's incomplete.
For example, if `getPipelineMap` fails, we can still show the result of `getContainerMap`. And print an error message to the user about the failed call.

Comment (siddhantsangwan) [inline hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DecommissionStatusSubCommand.java:87]: About `listPipelines` - since we're only talking about decommission here, we can avoid getting data about all pipelines over the network, and instead only get the pipelines with decommissioning nodes.

**Review by siddhantsangwan — COMMENTED** [review]
@Tejaskriya Thanks for working on this command, which can help us in understanding why datanode decommission is stuck.

I haven't gone over all the changes yet, but I have some initial thoughts about the amount of data we're transferring over the network and exception handling at the client side.

**Comment by siddhantsangwan** [comment]
> Decommission Status: DECOMMISSIONING - 1 nodes
> 
> Datanode: d616186b-fa8d-4353-8c75-f0eefa4279ba (/default-rack/172.19.0.8/ozone-datanode-4.ozone_default)
> Decommissioning started at : 1698730529868
> No. of pipelines when decommissioning was started: 3
> No. of pipelines currently: 0
> No. of containers when decommissioning was started: {QUASI_CLOSED=0, DELETING=0, OPEN=3, CLOSED=0, DELETED=0, CLOSING=0, RECOVERING=0}
> No. of containers currently: {QUASI_CLOSED=0, DELETING=0, OPEN=0, CLOSED=3, DELETED=0, CLOSING=0, RECOVERING=0}

The way the output is being displayed looks good to me. Just one question, what format is the time in?

**Comment by Tejaskriya** [comment]
> The way the output is being displayed looks good to me. Just one question, what format is the time in?

The time is in the form of milliseconds of the system time. i.e., System.currentTimeMillis()

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DecommissionStatusSubCommand.java (hunk 1 above)
Comment (Tejaskriya) [reply hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DecommissionStatusSubCommand.java:72]: As this is not a trivial change, I'll handle this in [HDDS-9648](https://issues.apache.org/jira/browse/HDDS-9648) APIs to fetch only one DN information will be introduced under it

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DecommissionStatusSubCommand.java (hunk 2 above)
Comment (Tejaskriya) [reply hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DecommissionStatusSubCommand.java:87]: This too will be handled under [HDDS-9648](https://issues.apache.org/jira/browse/HDDS-9648)

**Comment by sodonnel** [comment]
Just to be sure I am not missing something - this PR does not provide the complete functionality? For example, the containers pending replication are not yet included in the output?

We already have metrics in NodeDecommissionMetrics that keeps track of all nodes decommissioning, and also the various counts of under-replicated, open-pipelines etc. The only thing the metrics do not provide, is the list of containers in the degraded states.

I think the baseline information could therefore be provided by pulling the metrics from SCM (like the ozone insight commands do), and displaying it in a report form. Then we just need a way to get a list of the containers that are blocking decommission / maintenance by state, which could be obtained via a single new call into SCM.

We would not be able to get the starting container counts vs current counts, but these counts are not overly important unless some containers refuse to close - the counts at the start are not overly interesting.

I am not sure if the decommission start time is part of the metrics, but we could easily add it if it is not there already.

I think it is worth considering this approach, as it means we have a consistent view of metrics and the report output, and the numbers should align. I also think it will be much less code.

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DecommissionStatusSubCommand.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+import java.util.stream.Collectors;
+
+import static org.apache.hadoop.hdds.protocol.proto.HddsProtos.NodeOperationalState.DECOMMISSIONED;
+import static org.apache.hadoop.hdds.protocol.proto.HddsProtos.NodeOperationalState.DECOMMISSIONING;
+
+/**
+ * Handler to print decommissioning nodes status.
+ */
+@CommandLine.Command(
+    name = "decommissionStatus",
```
Comment (adoroszlai) [inline hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DecommissionStatusSubCommand.java:48]: Nit: please don't use `camelCase` for command (and option) names

Comment (Tejaskriya) [reply hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DecommissionStatusSubCommand.java]: Thank you for the review, I have changed it to decommission-status now

**Review by adoroszlai — COMMENTED** [review]
Thanks @Tejaskriya for updating the patch.  Can you please omit files with whitespace-only change?

```
hadoop-hdds/interface-admin/src/main/proto/ScmAdminProtocol.proto
hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/NodeDecommissionManager.java
hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/SimpleMockNodeManager.java
```

**Comment by Tejaskriya** [comment]
> Can you please omit files with whitespace-only change?

Thanks @adoroszlai , I have removed the whitespace-only changes now.

Currently this patch only introduces the framework for the command (only the information of datanodes in DECOMMISSIONING are displayed). It does not provide the complete functionality. The counts I had created earlier is removed as suggested because they are not important. 
The link to the jiras under which further functionalities will be handled is mentioned in the description of the PR.

@siddhantsangwan @sodonnel could you help me by reviewing this patch?

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DecommissionStatusSubCommand.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    for (HddsProtos.Node node : decommissioningNodes) {
+      DatanodeDetails datanode = DatanodeDetails.getFromProtoBuf(
+          node.getNodeID());
+      printDetails(datanode);
+    }
+  }
+  private void printDetails(DatanodeDetails datanode)
+      throws IOException {
```
Comment (siddhantsangwan) [inline hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DecommissionStatusSubCommand.java:76]: The method doesn't actually throw an exception.

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DecommissionStatusSubCommand.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+ */
+@CommandLine.Command(
+    name = "decommission-status",
+    description = "Show decommission status for datanodes",
+    mixinStandardHelpOptions = true,
+    versionProvider = HddsVersionProvider.class)
+
+public class DecommissionStatusSubCommand extends ScmSubcommand {
+
+  @CommandLine.Option(names = { "--id" },
```
Comment (siddhantsangwan) [inline hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DecommissionStatusSubCommand.java:45]: I suggest also having an option to get information by ip address. This can be done in future PRs if that's convenient.

**Review by siddhantsangwan — COMMENTED** [review]
@Tejaskriya Are you planning to support the `--json` option in this PR?

**Comment by Tejaskriya** [comment]
> @Tejaskriya Are you planning to support the `--json` option in this PR?

Thanks for the review, I plan to add and handle the `--json` option in [HDDS-9759](https://issues.apache.org/jira/browse/HDDS-9759)

**Comment by siddhantsangwan** [comment]
This change largely looks good. It'll be good to also have a test case for when the uuid the user provided is not decommissioning and the returned list is empty (For `"Datanode: " + uuid + " is not in DECOMMISSIONING"`).

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DecommissionStatusSubCommand.java (hunk 5 above)
Comment (Tejaskriya) [reply hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DecommissionStatusSubCommand.java:45]: I have pushed a commit now in which i have added the ipAddress option

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DecommissionStatusSubCommand.java [hunk 6]
```diff
@@ -59,6 +64,15 @@ public void execute(ScmClient scmClient) throws IOException {
         System.err.println("Datanode: " + uuid + " is not in DECOMMISSIONING");
         return;
       }
+    } else if (!Strings.isNullOrEmpty(ipAddress)) {
+      decommissioningNodes = scmClient.queryNode(DECOMMISSIONING, null,
+          HddsProtos.QueryScope.CLUSTER, "").stream().filter(p ->
+          p.getNodeID().getIpAddress().compareToIgnoreCase(ipAddress) == 0)
+          .collect(Collectors.toList());
+      if (decommissioningNodes.isEmpty()) {
+        System.err.println("Datanode: " + uuid + " is not in DECOMMISSIONING");
```
Comment (siddhantsangwan) [inline hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DecommissionStatusSubCommand.java:73]: `uuid` should be `ipAddress`.

File: hadoop-hdds/tools/src/test/java/org/apache/hadoop/hdds/scm/cli/datanode/TestDecommissionStatusSubCommand.java [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+ */
+public class TestDecommissionStatusSubCommand {
+
+  private static final String DEFAULT_ENCODING = StandardCharsets.UTF_8.name();
+  private final ByteArrayOutputStream outContent = new ByteArrayOutputStream();
+  private final ByteArrayOutputStream errContent = new ByteArrayOutputStream();
+  private final PrintStream originalOut = System.out;
+  private final PrintStream originalErr = System.err;
+  private DecommissionStatusSubCommand cmd;
+  private DecommissionSubCommand decomCmd;
```
Comment (siddhantsangwan) [inline hadoop-hdds/tools/src/test/java/org/apache/hadoop/hdds/scm/cli/datanode/TestDecommissionStatusSubCommand.java:58]: It looks like we can simply remove `decomCmd` from the tests since we're mocking `ScmClient`? If it's not necessary or doesn't do anything, let's remove it.

**Review by siddhantsangwan — COMMENTED** [review]
Thanks for the update. This should be good to merge once my latest comments are addressed.

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DecommissionStatusSubCommand.java [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        return;
+      }
+    } else if (!Strings.isNullOrEmpty(ipAddress)) {
+      decommissioningNodes = scmClient.queryNode(DECOMMISSIONING, null,
+          HddsProtos.QueryScope.CLUSTER, "").stream().filter(p ->
+          p.getNodeID().getIpAddress().compareToIgnoreCase(ipAddress) == 0)
+          .collect(Collectors.toList());
+      if (decommissioningNodes.isEmpty()) {
+        System.err.println("Datanode: " + uuid + " is not in DECOMMISSIONING");
+        return;
```
Comment (adoroszlai) [inline hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DecommissionStatusSubCommand.java:74]: We can reduce code duplication between the two branches.  Example for applying various filters to the same base stream:

https://github.com/apache/ozone/blob/5a9dbb85032cb3c5c6066010c8e9e91c680d6301/hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/ListInfoSubcommand.java#L85-L105

Comment (Tejaskriya) [reply hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DecommissionStatusSubCommand.java]: I have followed this approach now. Thank you for the review!

File: hadoop-hdds/tools/src/test/java/org/apache/hadoop/hdds/scm/cli/datanode/TestDecommissionStatusSubCommand.java (hunk 7 above)
Comment (Tejaskriya) [reply hadoop-hdds/tools/src/test/java/org/apache/hadoop/hdds/scm/cli/datanode/TestDecommissionStatusSubCommand.java]: I have removed it and used scmClient in the last commit pushed

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DecommissionStatusSubCommand.java [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static org.apache.hadoop.hdds.protocol.proto.HddsProtos.NodeOperationalState.DECOMMISSIONING;
+
+/**
+ * Handler to print decommissioning nodes status.
+ */
+@CommandLine.Command(
+    name = "decommission-status",
```
Comment (adoroszlai) [inline hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DecommissionStatusSubCommand.java:39]: On another look, I have concerns about adding a command that is a "continuation" of an existing command (`decomission`):

```
$ ozone admin datanode -h
Usage: ozone admin datanode [-hV] [COMMAND]
Datanode specific operations
  -h, --help      Show this help message and exit.
  -V, --version   Print version information and exit.
Commands:
  list                 List info of datanodes
  decommission         Decommission a datanode
  maintenance          Put a datanode into Maintenance Mode
  recommission         Return a datanode to service
  usageinfo            List usage information (such as Capacity, SCMUsed,
                         Remaining) of a datanode by IP address or UUID
  decommission-status  Show decommission status for datanodes
```

Some ideas:

* rename to `status` -- if it can be extended later to show other type of status info about datanodes
* rename to `status` and make it a subcommand of `decommission`; so the complete command would be `ozone admin datanode decommission status`

What do you think?

Comment (Tejaskriya) [reply hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DecommissionStatusSubCommand.java:39]: As we want to show the status of decommissioning nodes only in this status command, I think we can go ahead with the 2nd idea. I'll make it a subcommand of `decommission`

Comment (adoroszlai) [reply hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DecommissionStatusSubCommand.java:39]: Thanks @Tejaskriya, and sorry for bringing it up this late.

Comment (adoroszlai) [reply hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DecommissionStatusSubCommand.java:39]: One more idea: instead of `decommission status`, how about `status decommission`?  To avoid accidentally starting decommissioning of a node by an incomplete command...

Comment (Tejaskriya) [reply hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DecommissionStatusSubCommand.java:39]: Do we make it `status-decommission` as subcommand of `ozone admin datanode` instead?

Comment (adoroszlai) [reply hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DecommissionStatusSubCommand.java:39]: I prefer `status decommission`, because it is extensible.  While this command may be specific to decommissioning status, maybe in the future we'd like to add `status replication`, `status volumes`, or similar.

BTW, each subcommand can have aliases, so we could provide abbreviation as `ozone admin dn st decom`, if we want to save some typing (similar to `git st`, which is a commonly used alias for `git status`).  We can also do that with the single `status-decommission` subcommand, but somehow `st-decom` feels more arbitrary.

Comment (adoroszlai) [reply hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DecommissionStatusSubCommand.java]: Thanks, I'm happy with the latest version.

**Comment by adoroszlai** [comment]
@sodonnel would you like to take a look?
